aten/src/ATen/cuda/PhiloxRNGEngine.h

```diff
@@ -0,0 +1,284 @@
+      // if overflow in x has occured, carry over to nhi
+      unsigned int nhi_old = nhi;
+      nhi++;
+      counter.y += nhi;
+      if(nhi_old <= nhi) {
+        // if overflow in nhi has occured during carry over,
+        // propagate that overflow to y and exit to increment z
+        // otherwise return
+        if (nhi <= counter.y) {
+          return;
+        }
+      }
+    } else {
+      // if overflow in y has occured during addition,
+      // exit to increment z
+      // otherwise return
+      counter.y += nhi;
+      if (nhi <= counter.y) {
+        return;
+      }
+    }
+    if (++counter.z)
+      return;
+    ++counter.w;
+  }
+
+  /*
+  * Function that Skips one 128 bit number in a subsequence
+  */
+  C10_DEVICE inline void incr() {
+    if (++counter.x)
+      return;
+    if (++counter.y)
+      return;
+    if (++counter.z) {
+      return;
+    }
+    ++counter.w;
+  }
+
+private:
+  uint4 counter;
+  uint4 output;
+  uint2 key;
+  unsigned int STATE;
+
+  C10_DEVICE unsigned int mulhilo32(unsigned int a, unsigned int b,
+                                    unsigned int *result_high) {
+    *result_high = __umulhi(a, b);
+    return a*b;
+  }
+
+  C10_DEVICE inline uint4 single_round(uint4 ctr, uint2 key) {
+    unsigned int hi0;
+    unsigned int hi1;
+    unsigned int lo0 = mulhilo32(kPhiloxSA, ctr.x, &hi0);
+    unsigned int lo1 = mulhilo32(kPhiloxSB, ctr.z, &hi1);
+    uint4 ret = {hi1 ^ ctr.y ^ key.x, lo1, hi0 ^ ctr.w ^ key.y, lo0};
+    return ret;
+  }
+  static const unsigned long kPhilox10A = 0x9E3779B9;
+  static const unsigned long kPhilox10B = 0xBB67AE85;
+  static const unsigned long kPhiloxSA = 0xD2511F53;
+  static const unsigned long kPhiloxSB = 0xCD9E8D57;
+};
+
+typedef philox_engine Philox4_32_10;
+
+/*
+* Distribution implementations adapted from THRandom
+* Note: These functions are currently here as stubs and may be
+* removed in the future. There may be better ways of implementing 
+* them, especially for CUDA.
+*/
+
+/*
+* Produces a uniform distribution in the range [0,1) of type double
+* Note: how to get a range of [0,1) from {0,1,..2^64-1}
+* https://lemire.me/blog/2017/02/28/how-many-floating-point-numbers-are-in-the-interval-01/
+*/
+C10_DEVICE  __inline__ double standard_uniform_distribution(Philox4_32_10& engine) {
+  const uint32_t random32_val = engine();
+  double result = (random32_val & ((1ULL << 24) - 1)) * ::ldexp(1.0, -24);
+  return result;
+}
+
+/*
+* Produces a normal distribution given philox random, with mean = 0 and standard deviation = 1
+*/
+C10_DEVICE  __inline__ double normal_distribution(Philox4_32_10& engine) {
+  // Box-Muller method (https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform)
+  // We are not caching and just returning one of the variables.
+  double result;
+  double u2 = standard_uniform_distribution(engine);
+  double u1 = standard_uniform_distribution(engine);
+  double r = ::sqrt(-2.0 * ::log(u1));
+  result = r * ::cos(2.0 * M_PI * u2);
```

Comment:
ouch, double sqrt, log and cos will bring everything to a screeching halt.

Reply:
Made everything floats now.